core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -183,6 +194,15 @@ public Ec2MetadataAsyncBuilder endpointMode(EndpointMode endpointMode) {
         @Override
         public Ec2MetadataAsyncBuilder httpClient(SdkAsyncHttpClient httpClient) {
             this.httpClient = httpClient;
+            this.httpClientIsInternal = false;
+            return this;
+        }
+
+        @Override
+        public Builder httpClient(DefaultSdkAsyncHttpClientBuilder builder) {
+            SdkAsyncHttpClient builtHttpClient = builder.build();
+            this.httpClientIsInternal = true;
+            this.httpClient = builtHttpClient;
```

Comment:
Can we move this logic to the ctor of `DefaultEc2MetaAsyncClient`? Builder should be as simple as possible IMO.

Reply:
If users call `httpClientBuilder(xxx)` multiple times, we'd create multiple HTTP instances, which could cause memory leak in extreme cases...
```
Ec2MetaAsyncClient.builder()
                  .httpClientBuilder(xxx)
                  .httpClientBuilder(xxx)
                  ...
                  .build()
```

Reply:
It seems we can achieve the same in the ctor. IMO builder class should be treated as a simple POJO class w/o too much logic. We should probably add validation as well to make sure only one of the setters is called. 
https://github.com/aws/aws-sdk-java-v2/blob/2d27e75f6241913e028170b79c7e7072ce06d008/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java#L343-L349